### ISSUE TITLE
Update Contact UDFS

### ISSUE THREAD

**Issue by MadRainbow** [ecitsolutions/Autotask#78]
This has been driving me nuts for the whole day, and I am probably missing something really obvious.

How do i update UDF's in the Contact entity?

I can update all other fields, with 

 _set-atwscontact -Id XXXXX -Firstname "Test"_

however, I can't work out of the life of me how to update a UDF.  

I can call the UDF's with GET-ATWSCONTACT, however using the returned variable name in a SET simply throws up an error.....

_Set-AtwsContact : A positional parameter cannot be found that accepts argument '-#Authentication Mobile'.
At line:1 char:1_
  
Any thoughts?

 Thanks

David

**Comment by JsonWud** [comment]
Hi David,

The properties of entities/objects that start with "#" are really just
read-only copies of the UDFs that have been hoisted up to the same level as
the rest of the entity/object properties and cannot be used to update the
UDF directly. Think of this as a convenience to make comparisons and
filtering easier.

I just whipped this up to help you understand how it's done with the real
UDFs. I hope it is helpful.

# There is not a single way to work with UserDefinedFields. Below is
the way I usually interact with them.

# Understanding the shape of UserDefinedFields is usually one of the
largest hurdles to becoming familiar
# with the Autotask API. Note the example below starts at the
UserDefinedField level, which is a property
# of the entity/object you are targeting. The UDF values are actually
3 levels below that of the entity.
# Contact Entity -> UDF Property -> Array of Hashtables -> Values
inside the Hashtable

$UserDefinedFields = @{Name = 'Shirt Color'; Value = 'Blue' },
# Structure is an array of Hashtables.
                     @{Name = 'Shirt Size'; Value = 'Medium' },
# If this looked like @{'Shirt Size' = 'Medium'}
                     @{Name = 'Shirt Material'; Value = 'Polyester' },
# it would be easier to reference the element
                     @{Name = 'Warranty Expiration'; Value = '1/1/2021' }
# directly by key.

# If I need to update the value of Shirt Size, I have to find the
element of the array where the value of Name is 'Shirt Size'.
# Below is one method to achieve this:

($UserDefinedFields | Where-Object { $_.Name -eq 'Shirt Size' }).Value =
'Large'

# Now I can use the 'Set' command for this entity to update that value
in Autotask.

I would suggest spending a little time pulling in entities into your
powershell session and Looking at UDFs using the
$MyEntity.UserDefinedFields syntax, looking at the type and members.

Feel free to follow up with any other questions. Perhaps others will chime
in with alternate methods.

Thanks,
Jason

**Comment by MadRainbow** [comment]
Hi Jason,

Thanks for the quick response.

Not sure why this isn't making sense to me (i've spent twenty odd years programming), or if this is a bug I am working against.

Pulling the UDF's into the PS session isn't a a problem, can grab those as needed.  The bit that's driving me up the wall (and lets face it, whilst we're in isolation, that ain't a good thing!)

So, i have run the command

_set-atwscontact -id 30684442 -UserDefinedFields @{name='Authentication Mobile'; value='123'}_

And get the response

_Set-AtwsContact : Parameter set cannot be resolved using the specified named parameters._

Am I still not grasping this, or could this be a bug?

**Comment by JsonWud** [comment]
Hi David,

TLDR - Not a bug, but incorrect command usage :)

Sorry you're having a tough time with this. I see where your error is coming from. In the command above (set-atwscontact -id 30684442 -UserDefinedFields @{name='Authentication Mobile'; value='123'}) you are actually mixing parameter sets. The ID parameter is only a member of the 'By_Id' parameterset and the UserDefinedFields parameter is a member of the 'Input_Object' and 'By_parameters' parametersets. These two parameters cannot be used in the same command. This info can be found in the comment based help for the Set-AtwsContact command. (See attachments)

![ID_param_definition](https://user-images.githubusercontent.com/39089884/78505421-663bdc00-7741-11ea-9fce-a90dea14131a.png)
![UDF_param_definition](https://user-images.githubusercontent.com/39089884/78505422-66d47280-7741-11ea-9019-98539896b68d.png)

I'm personally very fond of pulling in the entire contact object, modifying the properties/UDFs in memory, then simply updating the entire thing using a command like this:
$myUpdate = Set-AtwsContact -InputObject $myModifiedContact -Passthru

This way I don't have to worry about parameter set conflicts. But again, there are a few ways to accomplish the update, and I don't want to say that mine is THE way.

You could rewrite your command so that it doesn't have the parameterset conflict like so:
Set-AtwsContact -EMailAddress 'target.user@domain.com' -UDF @{name='Authentication Mobile'; value='123'

Hope this is helpful and gets you moving forward on your task. Stay safe.
Jason

**Comment by MadRainbow** [comment]
I can't believe that I have spent the whole day trying to work this out....

I was clearly trying to update the ID number, when I wanted the -ID to be the WHERE

get-atwscontact -id 30684442 | Set-AtwsContact -UserDefinedFields @{name='Authentication Mobile'; value='1234'}

This worked!!!

Thanks for your help Jason....sorry for being so dense!

David
